File: app/src/main/java/com/example/java_viewmodel/mainlist/ItemRecyclerViewAdapter.java
package com.example.java_viewmodel.mainlist;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.java_viewmodel.R;

import java.util.Locale;
import java.util.Objects;

/**
 * {@link RecyclerView.Adapter}
 */
public class ItemRecyclerViewAdapter extends ListAdapter<ListItemValue, ItemRecyclerViewAdapter.ViewHolder> {

    private final LifecycleOwner viewLifecycleOwner;

    private static DiffUtil.ItemCallback<ListItemValue> getItemCallback() {
        return new DiffUtil.ItemCallback<ListItemValue>() {
            @Override
            public boolean areItemsTheSame(@NonNull ListItemValue oldItem, @NonNull ListItemValue newItem) {
                return oldItem == newItem;
            }

            @Override
            public boolean areContentsTheSame(@NonNull ListItemValue oldItem, @NonNull ListItemValue newItem) {
                return Objects.equals(oldItem.getNumber(), newItem.getNumber());
            }
        };
    }

    public ItemRecyclerViewAdapter(LifecycleOwner viewLifecycleOwner) {
        super(getItemCallback());
        this.viewLifecycleOwner = viewLifecycleOwner;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_list_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.mItem = getItem(position);
        holder.mNameView.setText(getItem(position).getName());
        getItem(position).getCurrentCountLiveData().observe(viewLifecycleOwner, holder.observer);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public final View mView;
        public final TextView mNameView;
        public final TextView mCountView;
        public ListItemValue mItem;

        private final Observer<Long> observer = new Observer<Long>() {
            @Override
            public void onChanged(Long aLong) {
                mCountView.setText(String.format(Locale.getDefault(), "%d", aLong));
            }
        };

        public ViewHolder(View view) {
            super(view);
            mView = view;
            mNameView = view.findViewById(R.id.item_name);
            mCountView = view.findViewById(R.id.item_count);
        }

        @NonNull
        @Override
        public String toString() {
            return super.toString() + " '" + mCountView.getText() + "'";
        }
    }
}

File: app/src/main/java/com/example/kt_viewmodel/mainlist/ItemRecyclerViewAdapter.kt
package com.example.kt_viewmodel.mainlist

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.kt_viewmodel.R
import java.util.*

/**
 * [RecyclerView.Adapter]
 */
class ItemRecyclerViewAdapter(private val viewLifecycleOwner: LifecycleOwner) : ListAdapter<ListItemValue, ItemRecyclerViewAdapter.ViewHolder>(itemCallback) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_list_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.mItem = getItem(position)
        holder.mNameView.text = getItem(position)?.name
        getItem(position)?.getCurrentCountLiveData()?.observe(viewLifecycleOwner, holder.observer)
    }

    class ViewHolder(mView: View) : RecyclerView.ViewHolder(mView) {
        val mNameView: TextView = mView.findViewById(R.id.item_name)
        private val mCountView: TextView = mView.findViewById(R.id.item_count)
        var mItem: ListItemValue? = null
        val observer: Observer<Long> = Observer { aLong -> mCountView.text = String.format(Locale.getDefault(), "%d", aLong) }
        override fun toString(): String {
            return super.toString() + " '" + mCountView.text + "'"
        }
    }

    companion object {
        private val itemCallback: DiffUtil.ItemCallback<ListItemValue>
            get() = object : DiffUtil.ItemCallback<ListItemValue>() {
                override fun areItemsTheSame(oldItem: ListItemValue, newItem: ListItemValue): Boolean {
                    return oldItem == newItem
                }

                override fun areContentsTheSame(oldItem: ListItemValue, newItem: ListItemValue): Boolean {
                    return oldItem.number == newItem.number
                }
            }
    }
}